src/paradigma/heart_rate/heart_rate_analysis.py

```diff
@@ -64,25 +65,84 @@ def signal_quality_classification(df: pd.DataFrame, config: SignalQualityClassif
+    df_windowed = tsdf.load_dataframe_from_binaries([metadata_time, metadata_values], tsdf.constants.ConcatenationType.columns)
+
+    df_sqa = signal_quality_classification(df_windowed, config, path_to_classifier_input)
+
+
+def estimate_heart_rate(df_sqa: pd.DataFrame, df_ppg_preprocessed: pd.DataFrame, config: HeartRateExtractionConfig) -> pd.DataFrame:  
+    """
+    Estimate the heart rate from the PPG signal using the time-frequency domain method.
+
+    Parameters
+    ----------
+    df_sqa : pd.DataFrame
+        The DataFrame containing the signal quality assessment predictions.
+    df_ppg_preprocessed : pd.DataFrame
+        The DataFrame containing the preprocessed PPG signal.
+    config : HeartRateExtractionConfig
+        The configuration for the heart rate estimation.
+
+    Returns
+    -------
+    df_hr : pd.DataFrame
+        The DataFrame containing the heart rate estimations.
+    """
+       
+    # Assign window-level probabilities to individual samples
+    ppg_post_prob = df_sqa.loc[:, DataColumns.PRED_SQA_PROBA].to_numpy()
+    #acc_label = df_sqa.loc[:, DataColumns.ACCELEROMETER_LABEL].to_numpy() # Adjust later in data columns to get the correct label, should be first intergrated in feature extraction and classification
+
+    sqa_label = assign_sqa_label(ppg_post_prob, config) # assigns a signal quality label to every individual data point
+    v_start_idx, v_end_idx = extract_hr_segments(sqa_label, config.min_hr_samples) # extracts heart rate segments based on the SQA label
+    
+    fs = config.sampling_frequency
+
+    v_hr_rel = np.array([])
+    t_hr_rel = np.array([])
+
+    edge_add = 2 * fs  # Add 2s on both sides of the segment for HR estimation
+
+    for start_idx, end_idx in zip(v_start_idx, v_end_idx):
+        # Skip if the epoch cannot be extended by 2s (edge_add) on both sides
+        if start_idx < edge_add or end_idx > len(df_ppg_preprocessed) - edge_add:
+            continue
+        
+        # Extract the extended PPG segment for HR estimation
+        extended_ppg_segment = df_ppg_preprocessed[DataColumns.PPG][start_idx - edge_add : end_idx + edge_add]
+
+        # Perform HR estimation
+        hr_est = extract_hr_from_segment(extended_ppg_segment, config.tfd_length, fs, config.kern_type, config.kern_params)
+
+        # Generate HR estimation time array
+        rel_segment_time = df_ppg_preprocessed[DataColumns.TIME][start_idx:end_idx].values # relative time in seconds after the start of the segment of each sample
+        n_full_segments = len(rel_segment_time) // config.hr_est_samples # number of full segments
+        hr_time = rel_segment_time[:n_full_segments * config.hr_est_samples : config.hr_est_samples] # relative time extracted from the full segments of every 2 seconds HR estimation
+
+        # Concatenate HR estimations and times
+        v_hr_rel = np.concatenate((v_hr_rel, hr_est))
```

Comment:
Concatenation of arrays can be inefficient when working with many loops. There is a smarter way by creating an empty array using `np.empty()` and setting the size beforehand in the empty call. Then, in the loop, you can add items to specific indices.